File: crates/hyper_rhi/src/render_pass.rs
//
// Copyright (c) 2024, SkillerRaptor
//
// SPDX-License-Identifier: MIT
//

use std::{fmt::Debug, sync::Arc};

use downcast_rs::Downcast;

use crate::{buffer::Buffer, graphics_pipeline::GraphicsPipeline, texture::Texture};

#[derive(Clone, Copy, Debug)]
pub struct DrawIndexedArguments {
    pub index_count: u32,
    pub instance_count: u32,
    pub first_index: u32,
    pub vertex_offset: i32,
    pub first_instance: u32,
}

#[derive(Clone, Debug)]
pub struct RenderPassDescriptor<'a> {
    pub label: Option<&'a str>,
    pub color_attachment: &'a Arc<dyn Texture>,
    pub depth_stencil_attachment: Option<&'a Arc<dyn Texture>>,
}

pub trait RenderPass: Downcast {
    fn set_pipeline(&self, pipeline: &Arc<dyn GraphicsPipeline>);

    fn set_index_buffer(&self, buffer: &Arc<dyn Buffer>);

    fn set_push_constants(&self, data: &[u8]);

    fn draw_indexed(&self, arguments: &DrawIndexedArguments);
}

downcast_rs::impl_downcast!(RenderPass);

File: hyper_engine/src/rhi/render_pass.cpp
/*
 * Copyright (c) 2025-present, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/render_pass.hpp"

#include "hyper_rhi/texture_view.hpp"

namespace hyper_engine
{
    RenderPass::RenderPass(const RenderPassDescriptor &descriptor)
        : m_label(descriptor.label)
        , m_label_color(descriptor.label_color)
        , m_color_attachments(descriptor.color_attachments)
        , m_depth_stencil_attachment(descriptor.depth_stencil_attachment)
    {
    }

    std::string_view RenderPass::label() const
    {
        return m_label;
    }

    LabelColor RenderPass::label_color() const
    {
        return m_label_color;
    }

    const std::vector<ColorAttachment> &RenderPass::color_attachments() const
    {
        return m_color_attachments;
    }

    DepthStencilAttachment RenderPass::depth_stencil_attachment() const
    {
        return m_depth_stencil_attachment;
    }
} // namespace hyper_engine